doc/dev/osd_internals/manifest.rst

```diff
@@ -279,6 +406,8 @@ Operations:
   TODO: For a redirect manifest, I don't see where it unsets the redirect.
         Is it supposed to?
 
+  --> need to invoke unset_manifest() (this is a rados api)
```

Comment:
Oh, I think the "Returns EINVAL if the object already is a redirect." is wrong. (right?)

So, if I'm understanding this correctly, invoking tier_promote and then failing to invoke unset_manifest would result in an object with copied data but a redirect still set.  I think that's really bad -- reads and writes would still go to the redirect target, and the osd's local data would essentially be leaked space.  I think you need to ensure that an object is redirected, or has contents, but never both.

Reply:
Unset_manifest will delete manifest info even if it has a redirect.
And, Redirect just forwards all operations to the redirect target. In the redirect object, tier_promote is meaningless . Do you mean SET_CHUNK? If so,  unset_manifest() will trigger both deleting contents and dropping the reference of a chunk.

Reply:
unset_manifest is a rados operation, right?  A user would have to actually send it?  What would happen if they perform tier_promote on a redirect object and then never send unset_manifest?

Reply:
Oh, right. I thought that I made the redirect to disallow to use tier_promote. 
How about disable using tier_promote in the case of the redirect?

Reply:
I had understood that a core component of your design was the ability to atomically replace a redirect or dedup'd chunk with a local copy atomically without clients noticing.  Wouldn't disallowing tier_promote here break that goal?

Reply:
In a redirect target, the basic concept is set_redirect and should use unset_manifest if the user want to reconfigure the redirect manifest object (do not allow to use redirect twice).
And, the redirect manifest object forwards all operation from the base tier to the lower tier (like proxy). So, Is there any problem if disallowing tier_promote? Sorry, I don't understand what you say.

Reply:
Ok.  Suppose you are writing an external tiering agent using these interfaces where tier_promote cannot be used on a redirect object.
1) You want to replace the cold blocks of an rbd image with redirects to a slower pool.
2) Later, you expect the image to become hot and want to replace those redirects with data resident in the base pool.
How do you use the redirect apis you've outlined to accomplish those two things?  Note: there might be racing normal reads and writes to the rbd image from a vm and your operations cannot change the answers those operations get.

Reply:
To do that,

1. Set_chunk an object (entire size) to the slower pool (note: set_chunk is not set_redirect)
2. become hot
3. invoke tier_promote to the base pool
4. write and read the local data

If racing reads and writes happen to the rbd imge between 1 and 3,
tier_promote block incoming reads and writes until it complete tier_promote.
After that, the pending reads and writes are re-queued.
Also, tier_promote is called after ongoing reads and writes are completed.

And, if the base pool means another pool,
1. Set_chunk an object A (entire size) to the slower pool
2. become hot
3. invoke tier_promote to promote the object A in the local pool
4. reads and writes will go to the local data
5. unset_manifest the object A ( but local data is still alive and can handle reads and writes)
6. Set_chunk the object A in the pool X (pool X is another pool)

Reply:
Well, this is a bit different as in order to do writes in your example you'd have to promote first, right?  Where do you see set_redirect being used?

Reply:
>Well, this is a bit different as in order to do writes in your example you'd have to promote first, right? 

I think there are two user cases.
The first case is:
1. Create Object A and B
2. Setchunk A to B
3. Write A
4. Flush occurs (deduped)

The second case is:
1. Create Object A and B
2. Setchunk A to B
3. Promote A
4. Write A
5. Flush occurs (deduped) 

The first case ignores the existing content. Otherwise, the second case promotes the existing content and use it.



> Where do you see set_redirect being used?

As far as I know, the purpose of set_redirect is two.
1. Redirect all operation to the target object (like proxy)
2. Read-proxy mode (weak consistency)

Reply:
Right, I'm saying that redirect offers something set_chunk does not, and that you'd need something like tier_promote to atomically replace a redirect with a copy in the base pool.  Thus, rather than disallowing tier_promote on redirect, you could just fix it by clearing the redirect as part of the promote.

Reply:
I got it.